tests/test_async_retry.py

```diff
@@ -0,0 +1,121 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2016-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import asyncio
+
+import logging
+
+from edgedb import _testbase as tb
+
+log = logging.getLogger(__name__)
+
+
+class Barrier:
+    def __init__(self, number):
+        self._counter = number
+        self._cond = asyncio.Condition()
+
+    async def ready(self):
+        if self._counter == 0:
+            return
+        async with self._cond:
+            self._counter -= 1
+            assert self._counter >= 0, self._counter
+            if self._counter == 0:
+                self._cond.notify_all()
+            else:
+                await self._cond.wait_for(lambda: self._counter == 0)
+
+
+class TestAsyncRetry(tb.AsyncQueryTestCase):
+
+    ISOLATED_METHODS = False
+
+    SETUP = '''
+        CREATE TYPE test::Counter EXTENDING std::Object {
+            CREATE PROPERTY name -> std::str {
+                CREATE CONSTRAINT std::exclusive;
+            };
+            CREATE PROPERTY value -> std::int32 {
+                SET default := 0;
+            };
+        };
+    '''
+
+    TEARDOWN = '''
+        DROP TYPE test::Counter;
+    '''
+
+    async def test_async_retry_01(self):
+        async for tx in self.con.retry():
+            async with tx:
+                await tx.execute('''
+                    INSERT test::Counter {
+                        name := 'counter1'
+                    };
+                ''')
+
+    async def test_async_retry_conflict(self):
+        con2 = await self.connect(database=self.get_database_name())
+        self.addCleanup(con2.aclose)
+
+        barrier = Barrier(2)
+        iterations = 0
+
+        async def transaction1(con):
+            async for tx in con.retry():
+                nonlocal iterations
+                iterations += 1
+                async with tx:
+                    # This magic query makes the test more reliable for some
+                    # reason. I guess this is because starting a transaction
+                    # in EdgeDB (and/or Postgres) is accomplished somewhat
+                    # lazily, i.e. only start transaction on the first query
+                    # rather than on the `START TRANSACTION`.
```

Comment:
I don't think this is the reason, the snapshot should be captured as soon as you start the transaction. So 🤷‍♂️

Reply:
Yes. Also what actually happens, is that second query is blocked rather than throwing a transaction error.

Also it looks like this also doesn't make the test 100% reliable :( (https://github.com/edgedb/edgedb-python/pull/146/checks?check_run_id=1709638747#step:7:1023) @elprans do you have any idea how to make this test reliable?

Reply:
> is that second query is blocked

Hm, that's weird. The current server doesn't have any deliberate functionality to block, and neither does Postgres (AFAIK).

Reply:
I think blocking is implied by `SELECT xxx FOR UPDATE`. So I assume it *might* apply to `ON DUPLICATE KEY` too.

Reply:
See below.